climada/util/lines_polys_handler.py

```diff
@@ -0,0 +1,886 @@
+    impact_pnt = Impact()
+    impact_pnt.calc(exp_pnt, impf_set, haz, save_mat=True)
+
+    return impact_pnt_agg(impact_pnt, exp_pnt, agg)
+
+
+def impact_pnt_agg(impact_pnt, exp_pnt, agg):
+    """
+    Aggregate the impact per geometry.
+
+    The output Impact object contains an extra attribute 'geom_exp'
+    containing the geometries.
+
+    Parameters
+    ----------
+    impact_pnt : Impact
+        Impact object with impact per exposure point (lines of exp_pnt)
+    exp_pnt : Exposures
+        Exposures with a double index geodataframe, first for geometries,
+        second for the point disaggregation of the geometries.
+    agg : string, optional
+        If 'agg', the impact is averaged over all points in each geometry.
+        If 'sum', the impact is summed over all points in each geometry.
+        The default is 'sum'.
+
+    Returns
+    -------
+    Impact
+        Impact object with the impact per geometry.
+
+    """
+
+    # aggregate impact
+    mat_agg = aggregate_impact_mat(impact_pnt, exp_pnt.gdf, agg)
+
+    #Write to impact obj
+    impact_agg = set_imp_mat(impact_pnt, mat_agg)
+
+    #Add exposure representation points as coordinates
+    repr_pnts = gpd.GeoSeries(
+        exp_pnt.gdf['geometry_orig'][:,0].apply(
+            lambda x: x.representative_point()
+            )
+        )
+    impact_agg.coord_exp = np.array([repr_pnts.y, repr_pnts.x]).transpose()
+    #Add geometries
+    impact_agg.geom_exp = exp_pnt.gdf.xs(0, level=1)\
+        .set_geometry('geometry_orig')\
+            .geometry.rename('geometry')
+
+    return impact_agg
+
+def aggregate_impact_mat(imp_pnt, gdf_pnt, agg):
+    """
+    Aggregate impact matrix given geodataframe or disaggregated polygons.
+
+    Parameters
+    ----------
+    impact_pnt : Impact
+        Impact object with impact per point (lines of gdf_pnt)
+    gdf_pnt : GeoDataFrame
+        Exposures geodataframe with a double index, first for polygon geometries,
+        second for the point disaggregation of the polygons.
+    agg : string
+        If 'agg', the impact is averaged over all points in each polygon.
+        If 'sum', the impact is summed over all points in each polygon.
+
+    Returns
+    -------
+    sparse.csr_matrix
+        matrix num_events x num_polygons with impacts.
+
+    """
+    # aggregate impact
+    mi = gdf_pnt.index
+    col_geom = mi.get_level_values(level=0).to_numpy()
+    row_pnt = np.arange(len(col_geom))
+    if agg == 'avg':
+        from collections import Counter
+        geom_sizes = Counter(col_geom).values()
+        mask = np.concatenate([np.ones(l) / l for l in geom_sizes])
+    elif agg == 'sum':
+        mask = np.ones(len(col_geom))
+    else:
+        raise ValueError("Please choose a valid aggregation method. agg=%s is not valid", agg)
+    csr_mask = sp.sparse.csr_matrix(
+        (mask, (row_pnt, col_geom)),
+         shape=(len(row_pnt), len(np.unique(col_geom)))
+        )
+    return imp_pnt.imp_mat.dot(csr_mask)
```

Comment:
Did you implement this for efficiency reasons like that? 
One could also do a one-liner using a multi-indexed pandas-df for the imp_mat and doing a df.groupby(outer level).sum / df.groupby(outer level).avg, instead. The advantage of that being readability. But perhaps it gets slow for _very_ large matrices.

Reply:
Yes. Whenever one is dealing with sparse matrices, one should never cast them into dense ones unless absolutely necessary. Also, linear algebra operations are very efficient. The code here is rather optimized and at least my first implementation along the lines you suggest was 3-4 orders of magnitude (1000 - 10000x) slower.

Reply:
Basically, if you can use a mask and a matrix multiplication instead of indexing/loops, always do so.